plutus-core/untyped-plutus-core/Language/UntypedPlutusCore/Evaluation/Machine/Cek.hs

```diff
@@ -0,0 +1,568 @@
+-- | The CEK machine.
+-- The CEK machine relies on variables having non-equal 'Unique's whenever they have non-equal
+-- string names. I.e. 'Unique's are used instead of string names. This is for efficiency reasons.
+-- The CEK machines handles name capture by design.
+-- Dynamic extensions to the set of built-ins are allowed.
+-- In case an unknown dynamic built-in is encountered, an 'UnknownDynamicBuiltinNameError' is returned
+-- (wrapped in 'OtherMachineError').
+
+{-# LANGUAGE ConstraintKinds       #-}
+{-# LANGUAGE DeriveAnyClass        #-}
+{-# LANGUAGE FlexibleInstances     #-}
+{-# LANGUAGE LambdaCase            #-}
+{-# LANGUAGE MultiParamTypeClasses #-}
+{-# LANGUAGE NPlusKPatterns        #-}
+{-# LANGUAGE OverloadedStrings     #-}
+{-# LANGUAGE RankNTypes            #-}
+{-# LANGUAGE TemplateHaskell       #-}
+{-# LANGUAGE TypeApplications      #-}
+{-# LANGUAGE TypeFamilies          #-}
+{-# LANGUAGE TypeOperators         #-}
+{-# LANGUAGE UndecidableInstances  #-}
+
+module Language.UntypedPlutusCore.Evaluation.Machine.Cek
+    ( CekValue(..)
+    , CekEvaluationException
+    , EvaluationResult(..)
+    , ErrorWithCause(..)
+    , EvaluationError(..)
+    , ExBudget(..)
+    , ExBudgetCategory(..)
+    , ExBudgetMode(..)
+    , ExRestrictingBudget(..)
+    , ExTally(..)
+    , CekExBudgetState
+    , CekExTally
+    , exBudgetStateTally
+    , extractEvaluationResult
+    , runCek
+    , runCekCounting
+    , evaluateCek
+    , unsafeEvaluateCek
+    , readKnownCek
+    )
+where
+
+import           PlutusPrelude
+
+import           Language.UntypedPlutusCore.Core
+import           Language.UntypedPlutusCore.Subst
+
+import           Language.PlutusCore.Constant
+import           Language.PlutusCore.Error
+import           Language.PlutusCore.Evaluation.Machine.ExBudgeting
+import           Language.PlutusCore.Evaluation.Machine.Exception
+import           Language.PlutusCore.Evaluation.Machine.ExMemory
+import           Language.PlutusCore.Evaluation.Result
+import           Language.PlutusCore.Name
+import           Language.PlutusCore.Pretty
+import           Language.PlutusCore.Universe
+
+import           Control.Lens                                       (AReview)
+import           Control.Lens.Operators
+import           Control.Lens.Setter
+import           Control.Monad.Except
+import           Control.Monad.Morph
+import           Control.Monad.Reader
+import           Control.Monad.State.Strict
+import           Data.Array
+import           Data.Hashable
+import           Data.HashMap.Monoidal
+import qualified Data.Map                                           as Map
+import           Data.Text.Prettyprint.Doc
+
+{- Note [Scoping]
+   The CEK machine does not rely on the global uniqueness condition, so the renamer pass is not a
+   prerequisite. The CEK machine correctly handles name shadowing.
+-}
+
+type TermWithMem uni = Term Name uni ExMemory
+
+{- Note [Arities in VBuiltin]
```

Comment:
I'd like to drop the typed CEK machine entirely, so that we don't need to update two machines every time we want to change something. So I don't want to spend time polishing docs of the typed machine.

Reply:
I think some of these comments (the one about arities, for example) are duplicated (or even triplicated!) in both the typed CEK machine and the typed CK machine.  When I changed to the new "value" types I added identical notes to both machines because I couldn't think of a more sensible thing to do.  Maybe cross-references to comments in other files would make sense, although then there's a danger that if you change a comment in one file it no longer makes sense in the context of a reference from a different file, and it would be easy to forget to change things in that other file.

Reply:
> Maybe cross-references to comments in other files would make sense, although then there's a danger that if you change a comment in one file it no longer makes sense in the context of a reference from a different file, and it would be easy to forget to change things in that other file.

Yes, docs tend to outdate pretty quickly. And cross-references would only make that problem more severe. I'd rather duplicate docs in this particular case. Or we can abstract out some of the code together with accompanying docs. Anyway, I want to finish off this PR, we can think about the docs situation later.

Reply:
Getting rid of the typed CEK machine would simplify things a bit. Simplifying comments maybe isn't the best argument for deleting code though!